src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -17,9 +18,62 @@ private Analyzer()
+            public static (SyntaxKind nodeToGenerate, VariableDeclaratorSyntax declaratorToRemoveOpt) Analyze(
+                SwitchStatementSyntax node,
+                SemanticModel semanticModel,
+                out bool shouldRemoveNextStatement)
             {
-                return new Analyzer().AnalyzeSwitchStatement(node, out shouldRemoveNextStatement);
+                var analyzer = new Analyzer();
+                var nodeToGenerate = analyzer.AnalyzeSwitchStatement(node, out shouldRemoveNextStatement);
+
+                if (nodeToGenerate == SyntaxKind.SimpleAssignmentExpression &&
+                    analyzer.TryGetVariableDeclaratorAndSymbol(semanticModel) is var (declarator, symbol))
+                {
+                    if (shouldRemoveNextStatement &&
+                        semanticModel.AnalyzeDataFlow(node.GetNextStatement()).DataFlowsIn.Contains(symbol))
+                    {
+                        // Bail out if data flows into the next statement that we want to move
+                        // For example:
+                        //
+                        //      string name = "";
+                        //      switch (index)
+                        //      {
+                        //          case 0: name = "0"; break;
+                        //          case 1: name = "1"; break;
+                        //      }
+                        //      throw new Exception(name);
+                        //
+                        return default;
+                    }
+
+                    var declaration = declarator.GetAncestor<StatementSyntax>();
+                    if (declaration.Parent == node.Parent && declarator.Initializer is null)
+                    {
+                        var beforeSwitch = node.GetPreviousStatement() is {} previousStatement
+                            ? semanticModel.AnalyzeDataFlow(declaration, previousStatement)
+                            : semanticModel.AnalyzeDataFlow(declaration);
+                        if (!beforeSwitch.WrittenInside.Contains(symbol))
+                        {
+                            // Move declarator only if it has no initializer and it's not used before switch
+                            return (nodeToGenerate, declaratorToRemoveOpt: declarator);
+                        }
+                    }
+                }
+
+                return (nodeToGenerate, declaratorToRemoveOpt: null);
+            }
+
+            private (VariableDeclaratorSyntax, ISymbol)? TryGetVariableDeclaratorAndSymbol(SemanticModel semanticModel)
+            {
+                if (_assignmentTargetOpt.IsKind(SyntaxKind.IdentifierName) &&
+                    semanticModel.GetSymbolInfo(_assignmentTargetOpt).Symbol is
+                        { Kind: SymbolKind.Local, DeclaringSyntaxReferences: { Length: 1 } syntaxRefs } symbol &&
+                    syntaxRefs[0].GetSyntax() is VariableDeclaratorSyntax declarator)
```

Comment:
this is unreadable to me :-/ 

i think there's such a thing as going too overboard on using patterns.

Reply:
really? this is nothing compared to what I'm planning =)
seriously, it's just (exactly) a couple of property patterns, I don't know how else you could use patterns at all

Reply:
it's simliar to the reason i ask people to break up large expressions that mix/match a ton of operators of different precedence.  i.e. lots of `||`s and `&&`s and `!`s.  Basically as you compress more and more into an expression, the more cluttered and difficult it can actually be to piece out what's going on.

>  I don't know how else you could use patterns at all

Easy, break out separate concepts.  For example, with the above, you could try things like:

```c#
if (!_assignmentTargetOpt.IsKind(SyntaxKind.IdentifierName))  return default;
if (!semanticModel.GetSymbolInfo(_assignmentTargetOpt).Symbol is { Kind: SymbolKind.Local, DeclaringSyntaxReferences: { Length: 1 } syntaxRefs } symbol) return default;

return syntaxRefs[0].GetSyntax() is VariableDeclaratorSyntax declarator ? (declarator, symbol) : default;
```

```

Reply:
Other things that can help break things down:

```c#
var symbol = semanticModel.GetSymbolInfo(_assignmentTargetOpt).Symbol;
```

Now pattern match on this.  

Consider and be careful with how many variables are being introduced and at which levels that is happening in a pattern.

For example, i would find it far easier to read ` { prop1: pattern1, prop2: pattern2, prop3: pattern3 } ` than ` { prop1: pattern, outerProp: { innerProp: { evenMoreProp: pattern }, innerProp2: pattern } } pattern `  etc.

Reply:
>break out separate concepts

Oh I thought you were opposing the recursive pattern itself. Ideally it would be
```cs
{ Kind: SymbolKind.Local, DeclaringSyntaxReferences: [var syntaxRef] } symbol
```
But we don't have that luxury yet. :)

Anyways, I don't think it's too bad in this particular case. but I can try to break to multiple ifs if you want.

Reply:
>```cs
>return syntaxRefs[0].GetSyntax() is VariableDeclaratorSyntax declarator ? (declarator, symbol) : default;
>```



can't use `default` here we want null and it won't be target typed, so:
```cs
return syntaxRefs[0].GetSyntax() is VariableDeclaratorSyntax declarator
    ? (declarator, symbol)
    : ((VariableDeclaratorSyntax, ISymbol)?)null;
```
I'll go with an if.